server/app/import.go

```diff
@@ -0,0 +1,172 @@
+package app
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"path"
+	"path/filepath"
+
+	"github.com/krolaw/zipstream"
+
+	"github.com/mattermost/focalboard/server/model"
+	"github.com/mattermost/focalboard/server/services/store"
+	"github.com/mattermost/focalboard/server/utils"
+
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+const (
+	archiveVersion = 2
+)
+
+// ImportArchive imports an archive containing zero or more boards, plus all
+// associated content, including cards, content blocks, views, and images.
+//
+// Archives are ZIP files containing a `version.json` file and zero or more
+// directories, each containing a `board.jsonl` and zero or more image files.
+func (a *App) ImportArchive(r io.Reader, opt model.ImportArchiveOptions) error {
+	zr := zipstream.NewReader(r)
+
+	boardMap := make(map[string]string) // maps old board ids to new
+
+	for {
+		hdr, err := zr.Next()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				a.logger.Debug("import archive - done", mlog.Int("boards_imported", len(boardMap)))
+				return nil
+			}
+			return err
+		}
+
+		dir, filename := path.Split(hdr.Name)
+		dir = path.Clean(dir)
+
+		switch filename {
+		case "version.json":
+			ver, errVer := parseVersionFile(zr)
+			if errVer != nil {
+				return errVer
+			}
+			if ver != archiveVersion {
+				return model.NewErrUnsupportedArchiveVersion(ver, archiveVersion)
+			}
+		case "board.jsonl":
+			boardID, err := a.ImportBoardJSONL(zr, opt)
+			if err != nil {
+				return fmt.Errorf("cannot import board %s: %w", dir, err)
+			}
+			boardMap[dir] = boardID
+		default:
+			// import file/image;  dir is the old board id
+			boardID, ok := boardMap[dir]
+			if !ok {
+				a.logger.Error("skipping orphan image in archive",
+					mlog.String("dir", dir),
+					mlog.String("filename", filename),
+				)
+				continue
+			}
+			// save file with original filename so it matches name in image block.
+			filePath := filepath.Join(opt.WorkspaceID, boardID, filename)
+			_, err := a.filesBackend.WriteFile(zr, filePath)
+			if err != nil {
+				return fmt.Errorf("cannot import file %s for board %s: %w", filename, dir, err)
+			}
+		}
+
+		a.logger.Trace("import archive file",
+			mlog.String("dir", dir),
+			mlog.String("filename", filename),
+		)
+	}
+}
+
+// ImportBoardJSONL imports a JSONL file containing blocks for one board. The resulting
+// board id is returned.
+func (a *App) ImportBoardJSONL(r io.Reader, opt model.ImportArchiveOptions) (string, error) {
+	// TODO: Stream this once `model.GenerateBlockIDs` can take a stream of blocks.
+	//       We don't want to load the whole file in memory, even though it's a single board.
+	blocks := make([]model.Block, 0, 10)
+	lineReader := bufio.NewReader(r)
+
+	userID := opt.ModifiedBy
+	if userID == model.SingleUser {
+		userID = ""
+	}
+	now := utils.GetMillis()
+
+	lineNum := 1
+	for {
```

Comment:
nit: This is loading everything in memory, would be great to be able to batch this changes (I know is hard because the GenerateIDs, but probably we can get rid of that function call and handle the new Ids generation directly here during the loop.

Reply:
When we get rid of `contentOrder` field in a card's `fields` column then it will be easy to do this.  I have a proposal for that we are debating.  Currently I only load a single board at a time which I figured was ok for now, until we can address the content order issue.  I agree this can and should be improved to not load a whole board, and we will.  Do you feel it has to happen this release?

Reply:
No, I don't think so, I something that we need to tackle eventually, especially if the size of the exports start growing and growing, but I think we are ok for now.